lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -604,26 +621,73 @@ void LowerLayersPass::runOnModuleBody(FModuleOp moduleOp,
-/// non-empty test bench directory.
-void LowerLayersPass::maybeSetOutputDir(Operation *op) {
-  if (!testBenchDir.empty() && dut)
-    op->setAttr("output_file", hw::OutputFileAttr::getAsDirectory(
-                                   op->getContext(), testBenchDir,
-                                   /*excludeFromFileList=*/true));
+static Operation *setOutputFile(Operation *op, hw::OutputFileAttr file) {
+  op->setAttr("output_file", file);
+  return op;
 }
 
 void LowerLayersPass::setOutputFile(Operation *op, const Twine &filename) {
-  if (!testBenchDir.empty() && dut) {
-    op->setAttr("output_file", hw::OutputFileAttr::getFromDirectoryAndFilename(
-                                   op->getContext(), testBenchDir, filename,
-                                   /*excludeFromFileList=*/true));
+  ::setOutputFile(
+      op, hw::OutputFileAttr::getFromFilename(op->getContext(), filename,
+                                              /*excludeFromFileList=*/true));
+}
 
-    return;
+/// For each layerblock under the given module, record the
+/// name of the generated/extracted module.
+void LowerLayersPass::recordExtractedModuleNames() {
+  auto circuit = getOperation();
+  CircuitNamespace ns(circuit);
+  for (auto &op : *circuit.getBodyBlock()) {
+    if (auto module = dyn_cast<FModuleOp>(op)) {
+      module->walk([&](LayerBlockOp layerBlock) {
+        auto name = moduleNameForLayer(module.getModuleName(),
+                                       layerBlock.getLayerName());
+        moduleNames.insert({layerBlock, ns.newName(name)});
+      });
+    }
+  }
+}
+
+void LowerLayersPass::recordLayerOutputDirs() {
+  using LayerOpIter = Block::op_iterator<LayerOp>;
+  using LayerOpRange = std::pair<LayerOpIter, LayerOpIter>;
+
+  auto idx = [](LayerOpRange &range) -> LayerOpIter & { return range.first; };
+  auto end = [](LayerOpRange &range) -> LayerOpIter & { return range.second; };
+
+  SmallVector<FlatSymbolRefAttr> path;
+  SmallVector<LayerOpRange> stack;
+
+  auto circuit = getOperation();
+  for (auto layer : circuit.getBodyBlock()->getOps<LayerOp>()) {
+    while (true) {
+      auto outputDir = layer->getAttrOfType<hw::OutputFileAttr>("output_file");
+      if (outputDir) {
+        auto name = SymbolRefAttr::get(&getContext(), layer.getSymName(), path);
+        layerOutputDirTable[name] = outputDir;
+      }
+
+      auto &body = layer.getBody().getBlocks().front();
+      if (!body.empty()) {
+        auto begin = body.op_begin<LayerOp>();
+        auto end = body.op_end<LayerOp>();
+        path.emplace_back(FlatSymbolRefAttr::get(layer.getSymNameAttr()));
+        stack.emplace_back(begin, end);
+      }
+
+      while (!stack.empty() && idx(stack.back()) == end(stack.back())) {
+        stack.pop_back();
+        path.pop_back();
+      }
+
+      if (stack.empty())
+        break;
+
+      assert(idx(stack.back()) != end(stack.back()));
+      layer = *idx(stack.back());
+      ++idx(stack.back());
```

Comment:
What is this logic doing?

Reply:
The logic here is meant to update the current layer, move the iterator forwards, and restart the loop. If I did it right...

Reply:
The logic here is meant to update the current layer, move the iterator forwards, and restart the loop. If I did it right...

Reply:
This looks like overkill. Can't this just lookup the layer by symbol and see if it has an output file attribute when it needs it? Alternatively, this could be cached, though isn't that just a walk over `LayerOp`?